lib/propolis/src/vmm/time.rs

```diff
@@ -0,0 +1,788 @@
+}
+
+impl VmTimeData {
+    pub fn wall_clock(&self) -> Duration {
+        Duration::new(self.hres_sec, self.hres_ns as u32)
+    }
+}
+
+impl From<bhyve_api::vdi_time_info_v1> for VmTimeData {
+    fn from(raw: bhyve_api::vdi_time_info_v1) -> Self {
+        Self {
+            guest_freq: raw.vt_guest_freq,
+            guest_tsc: raw.vt_guest_tsc,
+            hrtime: raw.vt_hrtime,
+            hres_sec: raw.vt_hres_sec,
+            hres_ns: raw.vt_hres_ns,
+            boot_hrtime: raw.vt_boot_hrtime,
+        }
+    }
+}
+impl From<VmTimeData> for bhyve_api::vdi_time_info_v1 {
+    fn from(info: VmTimeData) -> Self {
+        bhyve_api::vdi_time_info_v1 {
+            vt_guest_freq: info.guest_freq,
+            vt_guest_tsc: info.guest_tsc,
+            vt_hrtime: info.hrtime,
+            vt_hres_sec: info.hres_sec,
+            vt_hres_ns: info.hres_ns,
+            vt_boot_hrtime: info.boot_hrtime,
+        }
+    }
+}
+
+pub fn import_time_data(
+    hdl: &VmmHdl,
+    time_info: VmTimeData,
+) -> std::io::Result<()> {
+    let raw = bhyve_api::vdi_time_info_v1::from(time_info);
+    crate::vmm::data::write(hdl, -1, bhyve_api::VDC_VMM_TIME, 1, raw)?;
+
+    Ok(())
+}
+
+pub fn export_time_data(hdl: &VmmHdl) -> std::io::Result<VmTimeData> {
+    let time_info: bhyve_api::vdi_time_info_v1 =
+        crate::vmm::data::read(hdl, -1, bhyve_api::VDC_VMM_TIME, 1)?;
+
+    Ok(VmTimeData::from(time_info))
+}
+
+/// Returns the current host hrtime and wall clock time
+//
+// The current host hrtime and wall clock time are exposed via the VMM time data
+// interface.
+//
+// The kernel side of the interface disables interrupts while it takes the clock
+// readings; in the absence of a function to translate between the two clock
+// values, this is a best effort way to read the hrtime and wall clock times as
+// close to as possible at the same point in time. Thus fishing this data out of
+// the VMM time data read payload is strictly better than calling
+// clock_gettime(3c) twice from userspace.
+pub fn host_time_snapshot(hdl: &VmmHdl) -> std::io::Result<(i64, Duration)> {
+    let ti = export_time_data(hdl)?;
+    let wc = Duration::new(ti.hres_sec, ti.hres_ns as u32);
+    let hrt = ti.hrtime;
+
+    Ok((hrt, wc))
+}
+
+/// Given an input representation of guest time data on a source host, and a
+/// current host hrtime and wallclock time on the target host, output an
+/// "adjusted" view of the guest time data. This data can be imported to bhyve
+/// to allow guest time (namely, the guest TSC and its device timers) to allow
+/// the guest's sense of time to function properly on the target.
+//  See comments inline for more details about how we calculate a new guest TSC
+//  and boot_hrtime.
+pub fn adjust_time_data(
+    src: VmTimeData,
+    dst_hrt: i64,
+    dst_wc: Duration,
+) -> Result<(VmTimeData, VmTimeDataAdjustments), TimeAdjustError> {
+    // Basic validation: there is no reason system hrtime should be negative,
```

Comment:
Optional/nit: would it be valid to check this when the timestamp is fetched and convert it to an unsigned value there, or do we use it in signed form in other places?

Reply:
chatted about this offline -- I did use unsigned values for hrtime values that really should never be negative (such as uptime) in earlier versions of writing this code. I found that it was kind of a pain to go back and forth between signed and unsigned, since there are more constrained overflow checks for signed values (and thus I would have to convert any unsigned values back to signed anyway, or otherwise handle that for overflow checks). For my own sanity, I chose to keep all hrtime values signed throughout. Maybe there's a nicer way to write this code, but I would rather not do this for now.